### PR TITLE
Schema id check#173

### LINKED ISSUE #173: message.add should validate that schema_id exists
As the network, we do not want messages to be able to be added to non-existent schemas.

## Description
- Ensure that schema id is validated when creating a message.
- Options:
  - Schema Id should be less than the current `schema_count`.
  - Or Schema exists

## Acceptance Criteria
- [ ] When I am adding a new message, if I set the schema id to one that doesn't exist, the message should be rejected.
- [ ] Weights + benchmarks should be updated

### PR DESCRIPTION
**Pull request by saraswatpuneet**
# Goal
The goal of this PR is: As the network, we do not want messages to be able to be added to non-existent schemas.

Closes #173 

# Checklist
- [x] Implement a generic SchemaProvider Trait and implement for schema pallet
- [x] Add documents for newly implemented trait
- [x] Add unit test testing the negative route
- [x] Benchmarks added
- [x] Weights updated

### CODE REVIEW

File: pallets/messages/src/lib.rs [hunk 1]
```diff
@@ -178,6 +183,9 @@ pub mod pallet {
 		) -> DispatchResultWithPostInfo {
 			let provider_key = ensure_signed(origin)?;
 
+			let schema = T::SchemaProvider::get_schema_by_id(schema_id);
+			ensure!(schema.is_some(), Error::<T>::InvalidSchemaId);
+
```
Comment (enddynayn) [inline pallets/messages/src/lib.rs:188]: maybe move this one under the other ensure so that it does not read if the second one fails.

File: pallets/schemas/src/lib.rs [hunk 2]
```diff
@@ -287,3 +293,26 @@ pub mod pallet {
...
+	/// Since benchmarks are using regular runtime, we can not use mocking for this loosely bounded
+	/// pallet trait implementation. To be able to run benchmarks successfully for any other pallet
+	/// that has dependencies on this one, we would need to define msa accounts on those pallets'
+	/// benchmarks, but this will introduce direct dependencies between these pallets, which we
+	/// would like to avoid.
+	/// To successfully run benchmarks without adding dependencies between pallets we re-defined
+	/// this method to return schema checks requested by the benchmark to also be some.
+	#[cfg(feature = "runtime-benchmarks")]
+	fn get_schema_by_id(schema_id: SchemaId) -> Option<SchemaResponse> {
+		Some(SchemaResponse {
```
[the remark this answers is not in the document]
Comment (enddynayn) [reply pallets/schemas/src/lib.rs:311]: good idea.

**Review by enddynayn — APPROVED** [review]

**Comment by enddynayn** [comment]
💯 nice work!

File: pallets/schemas/src/lib.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (enddynayn) [reply pallets/schemas/src/lib.rs:311]: good idea.

[the remark this answers is not in the document]
Comment (saraswatpuneet) [reply pallets/schemas/src/lib.rs:311]: @aramikm  the issue the schema_id dont exist in storage so have to run register schema in benchmark for messages
what you think , hence this suggestion will fail

[the remark this answers is not in the document]
Comment (saraswatpuneet) [reply pallets/schemas/src/lib.rs:311]: One way to do it is to have another function to do db write in setup ( not sure that counts in calculating weights if so it's not going to work) then this read will come through
But it's a hacky way if possible

**Comment by aramikm** [comment]
You can just ignore the response if err and return the dummy value

**Comment by saraswatpuneet** [comment]
Haha sometimes default mode of thinking is just too complex , good stuff
